### ISSUE TITLE
Docker Swarm - service resize during an update can leave inconsistent versions 

### ISSUE THREAD

**Issue by sap1ens** [docker/docker#26617]
**Description**

If you try to resize Docker Swarm service during the update some versions can be inconsistent. Example: 

<img width="1039" alt="screen shot 2016-09-15 at 8 34 48 pm" src="https://cloud.githubusercontent.com/assets/260702/18572903/da2d71a8-7b84-11e6-8dcf-78c90daaf34a.png">

Here I originally had 25 containers with version 0.01 and I initiated an update to 0.1. During the update I resized a number of replicas to 10 and update stopped. Part of the containers have the required version, but most of them still have the old one. So update is not atomic. 

**Steps to reproduce the issue:**
1. Create a Docker Swarm service (I'm following the steps from https://bretfisher.github.io/orchestration-workshop-with-docker/)
2. Update the number of replicas: `docker service update worker --replicas 25`
3. Push the new version of the image. 
4. Trigger version update: `docker service update worker --image $DOCKER_REGISTRY/dockercoins_worker:v0.1`
5. Wait for 20-30 seconds. You see how some replicas are becoming updated. 
6. Resize service with `docker service update worker --replicas 10`.

**Describe the results you received:**

After the last step update stops and we can have inconsistent versions, look at the screenshot again. You assume that all replicas will be either upgraded or downgraded eventually, but nothing happens after 5 hours. 

**Describe the results you expected:**
Versions should be consistent - ideally all replicas should be updated with the new version.

**Additional information you deem important (e.g. issue happens only occasionally):**

.

**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 06:03:55 2016
 OS/Arch:      linux/amd64
 Experimental: true

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 06:03:55 2016
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 12
 Running: 5
 Paused: 0
 Stopped: 7
Images: 37
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /mnt/docker/aufs
 Backing Filesystem: extfs
 Dirs: 119
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge overlay host null
Swarm: active
 NodeID: aysj9vj6qbimq99ls3wwvkyif
 Is Manager: true
 ClusterID: 3plidaaq0dkhwun004cyj5y56
 Managers: 5
 Nodes: 5
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 172.31.36.35
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.2.0-36-generic
Operating System: Ubuntu 15.10
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.675 GiB
Name: ip-172-31-36-35
ID: AJ2G:FYX6:CWQS:UTOM:4LVR:74M2:R24C:GVPS:PIUW:MQBQ:FXN5:SYCA
Docker Root Dir: /mnt/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: true
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

AWS.

**Comment by allencloud** [comment]
@sap1ens 
I experienced this situation. If we wait for the first update action to finish, the result will be consistent. While if we start to send the second update request before first one finishes, inconsistency happens.

I think this is not as good as what I thin before.

In addition, I think if we combine two update commands into a single one, it will work.

**Comment by sap1ens** [comment]
Yeah, imagine if this bug is not fixed and we have some sort of management UI for Swarm. In this case it's totally possible for one person starting an update and for someone else resizing the service at the same time. 

So we have to either add additional check to UI (which creates more coupling) or just return an error when we attempt to resize and version update is not finished. I can take a look at this scenario if it sounds ok.

Or we just fix the root cause :)

**Comment by cpuguy83** [comment]
ping @aaronlehmann

**Comment by nishanttotla** [comment]
@sap1ens thanks for the report! I tried the steps here a few times before finally hitting this on my machine with Docker 1.12.1

**Comment by aaronlehmann** [comment]
I believe this is fixed on master by https://github.com/docker/swarmkit/pull/1497

I'll flag this for backport to 1.12.2.

**Comment by lbguilherme** [comment]
This seems to also have been fixed by #26765

**Comment by cpuguy83** [comment]
Thanks @lbguilherme!
